### PR TITLE
Support for PITR

### LINKED ISSUE #336: Support for Point In Time Recovery
**Description**
Support for Point In Time Recovery (PITR) should be provided by Service Fabrik
1. SF would deleted backups older than retention period only if on last day of retention period a successful backup is present.
2. SF should be able to restore a service instance supporting PITR on any times tamp in last x days. (x is the retention period in days).
3. Several enhancements are made in restore flow.

### PR DESCRIPTION
**Pull request by sauravmndl**
Fixes #336 
1. Keeping older backups until a successful backup present
2. Enhancements in restore flow
3. UTs
4. Retry-backups

### CODE REVIEW

File: data-access-layer/iaas/BackupStoreForServiceInstance.js [hunk 1]
```diff
@@ -49,30 +49,17 @@ class BackupStoreForServiceInstance extends BackupStore {
...
-        return _.lt(filenameobject.started_at, isoDate);
-      };
-    }
 
     if (service_id && instance_guid) {
       prefix += `/${service_id}.${instance_guid}`;
       if (backup_guid) {
         prefix += `.${backup_guid}`;
         message = `Backup '${backup_guid}' not found`;
-      } else if (time_stamp) {
```
[the remark this answers is not in the document]
Comment (sauravmndl) [reply data-access-layer/iaas/BackupStoreForServiceInstance.js]: This I intentionally removed as not used anywhere and it gives the latest backup behind the time_stamp which is not required by any function at this point of time. That's the reason removed. Following is the PR.

https://github.com/cloudfoundry-incubator/service-fabrik-broker/pull/335/commits/a5e632e5af51476105095a147e39cef86b0395d0

File: api-controllers/ServiceFabrikAdminController.js [hunk 2]
```diff
@@ -538,7 +538,7 @@ class ServiceFabrikAdminController extends FabrikBaseController {
         } else if (opts.backup_guid) {
           this.validateUuid(opts.backup_guid, 'Backup GUID');
         } else if (opts.time_stamp) {
-          this.validateDateString(opts.time_stamp);
+          this.validateRestoreTimeStamp(opts.time_stamp);
```
[the remark this answers is not in the document]
Comment (ketakig12) [reply api-controllers/ServiceFabrikAdminController.js]: Same for [below line](https://github.com/sauravmndl/service-fabrik-broker/blob/a5e632e5af51476105095a147e39cef86b0395d0/api-controllers/ServiceFabrikAdminController.js#L550).

[the remark this answers is not in the document]
Comment (ketakig12) [reply api-controllers/ServiceFabrikAdminController.js]: We do not have constraint of # of restores during last x days for OOB like for normal? If we have, we will need to call "validateRestoreHistory".

File: api-controllers/ServiceFabrikApiController.js [hunk 3]
```diff
@@ -414,7 +415,13 @@ class ServiceFabrikApiController extends FabrikBaseController {
         } else if (backupGuid) {
           return this.validateUuid(backupGuid, 'Backup GUID');
         } else if (timeStamp) {
-          return this.validateDateString(timeStamp);
```
Comment (ketakig12) [inline api-controllers/ServiceFabrikApiController.js]: I guess same comment as precedence of backup guid and timestamp applies here.

File: api-controllers/ServiceFabrikApiController.js [hunk 4]
```diff
@@ -427,16 +434,27 @@ class ServiceFabrikApiController extends FabrikBaseController {
...
+        return timeStamp ?
+          this.backupStore
+          .listBackupsOlderThan(backupFileOptions, new Date(timeStamp))
+          .then(oldBackupArray =>
+            _.sortBy(oldBackupArray, ['started_at']))
+          .then(sortedArray =>
+            _.findLast(sortedArray, backup => backup.state ===  CONST.OPERATION.SUCCEEDED))
+          .tap(successfulBackup => {
+            if (_.isEmpty(successfulBackup)) {
+              throw new NotFound(`No backup successful found for service instance '${instanceId}' before time_stamp ${new Date(timeStamp)} `);
```
Comment (ketakig12) [inline api-controllers/ServiceFabrikApiController.js]: Maybe error msg can be: "Cannot restore service instance as no successful backup found for ..."

File: api-controllers/ServiceFabrikApiController.js [hunk 5]
```diff
@@ -427,16 +434,27 @@ class ServiceFabrikApiController extends FabrikBaseController {
...
+            }
+          }) :
+          this.backupStore.getBackupFile(backupFileOptions);
       })
-      .catchThrow(NotFound, new UnprocessableEntity(`No backup with guid '${backupGuid}' found in this space`))
+      .catchThrow(NotFound, new UnprocessableEntity(`Cannot restore for guid/timeStamp '${timeStamp || backupGuid}' found in this space`))
       .tap(metadata => {
         if (metadata.state !== 'succeeded') {
```
Comment (ketakig12) [inline api-controllers/ServiceFabrikApiController.js]: In case of timestamp based restore, from [this block](https://github.com/sauravmndl/service-fabrik-broker/blob/a5e632e5af51476105095a147e39cef86b0395d0/api-controllers/ServiceFabrikApiController.js#L437-L448), where will we get metadata from?

File: api-controllers/FabrikBaseController.js [hunk 6]
```diff
@@ -111,12 +112,42 @@ class FabrikBaseController extends BaseController {
...
   }
 
+  validateRestoreHistory(options) {
+    return this.backupStore
+      .getRestoreFile(options)
+      .then(metdata => {
+        let dateArray = _.get(metdata, 'date_history');
+        if (!_.isEmpty(dateArray)) {
+          _.remove(dateArray, date => {
+            const olderDateTillRestoreAllowed = Date.now() - 1000 * 60 * 60 * 24 * config.backup.restore_history_days;
```
[the remark this answers is not in the document]
Comment (sauravmndl) [reply api-controllers/FabrikBaseController.js]: Renamed the variable. But whether we should allow user to restore in a moth more than 10 times when older backups are deleted as per deletion logic need to discuss.

File: broker/lib/fabrik/DirectorManager.js [hunk 7]
```diff
@@ -867,22 +867,66 @@ class DirectorManager extends BaseManager {
...
+  updateHistoryOfDates(arrayOfDates, isoDateToUpdate) {
+    let updatedHistory = arrayOfDates || [];
+    updatedHistory.push(isoDateToUpdate);
+    _.remove(updatedHistory, date => {
+      const twoMonthOlderDate = Date.now() - 1000 * 60 * 60 * 24 * 60;
+      return Date.parse(date) < twoMonthOlderDate;
+    });
+    return updatedHistory.sort();
+  }
+  scheduleBackup(opts) {
```
Comment (ketakig12) [inline broker/lib/fabrik/DirectorManager.js]: There is scheduleBackup function in DirectorInstance as well. Not sure if we can modify and use the same, but else it might be confusing.

File: api-controllers/ServiceFabrikAdminController.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (sauravmndl) [reply api-controllers/ServiceFabrikAdminController.js]: time_stamp takes preference. check below **then** blocks. For OOB we don't allow PITR.

[the remark this answers is not in the document]
Comment (sauravmndl) [reply api-controllers/ServiceFabrikAdminController.js]: Will remove the code in OOB flow.

File: api-controllers/ServiceFabrikApiController.js [hunk 8]
```diff
@@ -427,16 +434,27 @@ class ServiceFabrikApiController extends FabrikBaseController {
           backup_guid: backupGuid,
           tenant_id: tenantId
         };
-        return this.backupStore
-          .getBackupFile(backupFileOptions);
+        return timeStamp ?
```
[the remark this answers is not in the document]
Comment (sauravmndl) [reply api-controllers/ServiceFabrikApiController.js]: ok

File: api-controllers/ServiceFabrikApiController.js (hunk 3 above)
Comment (sauravmndl) [reply api-controllers/ServiceFabrikApiController.js]: done

File: api-controllers/ServiceFabrikApiController.js (hunk 5 above)
Comment (sauravmndl) [reply api-controllers/ServiceFabrikApiController.js]: .listBackupsOlderThan(backupFileOptions, new Date(timeStamp)) provides metadata.

File: jobs/ScheduleBackupJob.js [hunk 9]
```diff
@@ -82,6 +82,7 @@ class ScheduleBackupJob extends BaseJob {
     const options = _.omit(job.attrs.data, 'trigger', 'type');
     return backupStore
       .listBackupsOlderThan(options, config.backup.retention_period_in_days)
```
Comment (ketakig12) [inline jobs/ScheduleBackupJob.js]: I understood that we are now keeping base backups for retention_period + 1 days. How is it handled? 
Also, currently the transaction logs uploaded by service are not deleted as part of it, will that be part of another PR?

File: api-controllers/ServiceFabrikApiController.js [hunk 10]
```diff
@@ -446,7 +464,7 @@ class ServiceFabrikApiController extends FabrikBaseController {
           arguments: _.assign({
             backup: _.pick(metadata, 'type', 'secret')
           }, req.body, {
-            backup_guid: backupGuid || metadata.backup_guid
+            backup_guid: backupGuid || _.get(metadata, 'backup_guid')
```
[the remark this answers is not in the document]
Comment (sauravmndl) [reply api-controllers/ServiceFabrikApiController.js]: I need backup_guid to pass to agent so that they get the metadata of base backup.

File: broker/lib/fabrik/DirectorManager.js [hunk 11]
```diff
@@ -867,22 +867,66 @@ class DirectorManager extends BaseManager {
...
-                state: lastOperation.state,
-                logs: logs
-              })
-            );
+          return Promise.all([this.agent
+              .getRestoreLogs(agent_ip), this.backupStore
+              .getRestoreFile(options)
+            ])
+            .spread((logs, metadata) => {
+              restoreMetadata = metadata;
```
[the remark this answers is not in the document]
Comment (sauravmndl) [reply broker/lib/fabrik/DirectorManager.js]: used for another use-case. Later on didn't use, forgot to remove.

File: broker/lib/fabrik/DirectorManager.js [hunk 12]
```diff
@@ -867,22 +867,66 @@ class DirectorManager extends BaseManager {
...
+                  instance_id: options.instance_guid,
+                  afterXminute: 3
+                });
+              }
+            });
         }
       });
   }
 
+  updateHistoryOfDates(arrayOfDates, isoDateToUpdate) {
```
[the remark this answers is not in the document]
Comment (sauravmndl) [reply broker/lib/fabrik/DirectorManager.js]: Only dealing with updating the dates. Might not be specific to restore. Might be used for something else.

File: broker/lib/fabrik/DirectorManager.js [hunk 13]
```diff
@@ -867,22 +867,66 @@ class DirectorManager extends BaseManager {
...
+    let updatedHistory = arrayOfDates || [];
+    updatedHistory.push(isoDateToUpdate);
+    _.remove(updatedHistory, date => {
```
[the remark this answers is not in the document]
Comment (sauravmndl) [reply broker/lib/fabrik/DirectorManager.js]: the history will grow then infinitely until instance is deleted. So if a user does 10 restores in a month. There would be 240 entries in two years. So we keep that?

File: broker/lib/fabrik/DirectorManager.js (hunk 7 above)
Comment (sauravmndl) [reply broker/lib/fabrik/DirectorManager.js]: renamed to scheduleBackup

File: jobs/ScheduleBackupJob.js (hunk 9 above)
Comment (sauravmndl) [reply jobs/ScheduleBackupJob.js]: this PR would be updated for transaction log deletion.

File: jobs/ScheduleBackupJob.js [hunk 14]
```diff
@@ -135,6 +136,27 @@ class ScheduleBackupJob extends BaseJob {
       });
   }
 
+  static filterOldBackups(oldBackupArray) {
+    let filteredOldBackups = [];
+    /* oldBackupArray : This aray should contain all older backups
+    // including last retenion day's backup. E.g. if retention period
+    // is 15 days it would include all backups in (15th, 16th, 17th ...) */
+    if (typeof oldBackupArray !== 'undefined' && oldBackupArray.length > 0) {
```
[the remark this answers is not in the document]
Comment (sauravmndl) [reply jobs/ScheduleBackupJob.js]: done

File: api-controllers/ServiceFabrikApiController.js [hunk 15]
```diff
@@ -414,7 +415,13 @@ class ServiceFabrikApiController extends FabrikBaseController {
         } else if (backupGuid) {
           return this.validateUuid(backupGuid, 'Backup GUID');
         } else if (timeStamp) {
-          return this.validateDateString(timeStamp);
+          return Promise.try(() => this.validateRestoreTimeStamp(timeStamp))
```
[the remark this answers is not in the document]
Comment (sauravmndl) [reply api-controllers/ServiceFabrikApiController.js]: changed

File: data-access-layer/iaas/BackupStoreForServiceInstance.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (jagadish-kb) [reply data-access-layer/iaas/BackupStoreForServiceInstance.js]: OK

File: jobs/ScheduleBackupJob.js (hunk 9 above)
Comment (jagadish-kb) [reply jobs/ScheduleBackupJob.js]: Transaction log deletion?

Comment (jagadish-kb) [reply jobs/ScheduleBackupJob.js]: so should we not merge this PR then?

File: api-controllers/FabrikBaseController.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (jagadish-kb) [reply api-controllers/FabrikBaseController.js]: As discussed please atleast add a TODO on the need for having `restore_history_days` which to me seems redundant.

File: jobs/BnRStatusPollerJob.js [hunk 16]
```diff
@@ -147,7 +148,17 @@ class BnRStatusPollerJob extends BaseJob {
...
-        this.doPostFinishOperation(operationStatusResponse, operationName, instanceInfo) : operationStatusResponse
+        this.doPostFinishOperation(operationStatusResponse, operationName, instanceInfo)
+        .tap(() => {
+          if (operationStatusResponse.state === CONST.OPERATION.FAILED) {
+            const options = {
+              instance_id: instance_guid,
+              repeatInterval: utils.getCronWithIntervalAndAfterXminute(plan.service.backup_interval, CONST.SCHEDULE.RETRY_DELAY),
+              type: CONST.BACKUP.TYPE.ONLINE
+            };
+            return cf.serviceFabrikClient.scheduleBackup(options);
```
Comment (jagadish-kb) [inline jobs/BnRStatusPollerJob.js]: can you do this in a retry block here. Just in case you cannot reschedule, then base backup wont be trigerred.

Comment (jagadish-kb) [reply jobs/BnRStatusPollerJob.js]: also why is `serviceFabrikClient` added under cf module? it should be out on its own right? Its a data-access-layer on its own. This is not a review comment to you, but for SF 2.0.
@subhankarc @amitmalav  - comments?

File: api-controllers/ServiceFabrikApiController.js [hunk 17]
```diff
@@ -434,21 +434,24 @@ class ServiceFabrikApiController extends FabrikBaseController {
...
-            if (_.isEmpty(successfulBackup)) {
-              throw new NotFound(`No backup successful found for service instance '${instanceId}' before time_stamp ${new Date(timeStamp)} `);
-            }
-          }) :
-          this.backupStore.getBackupFile(backupFileOptions);
+        if (timeStamp) {
+          return this.backupStore
+            .listBackupsOlderThan(backupFileOptions, new Date(timeStamp))
+            .then(oldBackups =>
+              _.sortBy(oldBackups, ['started_at']))
```
Comment (jagadish-kb) [inline api-controllers/ServiceFabrikApiController.js]: unnecessary sort, `listBackupsOlderThan` already has sort in predicate on `started_at`. please check.

File: api-controllers/ServiceFabrikApiController.js [hunk 18]
```diff
@@ -434,21 +434,24 @@ class ServiceFabrikApiController extends FabrikBaseController {
...
+            .then(sortedOldBackups =>
+              _.findLast(sortedOldBackups, backup => backup.state ===  CONST.OPERATION.SUCCEEDED))
+            .tap(successfulBackup => {
+              if (_.isEmpty(successfulBackup)) {
+                logger.error(`No backup successful found for service instance '${instanceId}' before time_stamp ${new Date(timeStamp)}`);
+                throw new NotFound(`Cannot restore service instance '${instanceId}' as no successful backup found for time_stamp ${timeStamp}`);
```
Comment (jagadish-kb) [inline api-controllers/ServiceFabrikApiController.js]: Dont throw errors from `tap` - such errors wont bubble up. Please use tap only for logging, throw errors always from `then` block. I had given similar comments at couple of other places. Please ensure its taken care of.

File: jobs/BnRStatusPollerJob.js (hunk 16 above)
Comment (sauravmndl) [reply jobs/BnRStatusPollerJob.js]: Retry logic is implicit here. Reason is if this fails will be rescheduled after 30 mins. If fails again check would be done by BnRStatuspoller and  will be rescheduled after 30 mins.

Comment (subhankarc) [reply jobs/BnRStatusPollerJob.js]: @jagadish-kb makes sense to move it out of cf.

File: api-controllers/ServiceFabrikApiController.js (hunk 18 above)
Comment (sauravmndl) [reply api-controllers/ServiceFabrikApiController.js]: done

File: broker/lib/fabrik/DirectorManager.js [hunk 19]
```diff
@@ -876,52 +875,49 @@ class DirectorManager extends BaseManager {
...
 
-  updateHistoryOfDates(arrayOfDates, isoDateToUpdate) {
-    let updatedHistory = arrayOfDates || [];
-    updatedHistory.push(isoDateToUpdate);
-    _.remove(updatedHistory, date => {
-      const twoMonthOlderDate = Date.now() - 1000 * 60 * 60 * 24 * 60;
+  updateHistoryOfRestoreDates(restoreDates, isoDateToUpdate) {
+    restoreDates = restoreDates || [];
+    restoreDates.push(isoDateToUpdate);
+    _.remove(restoreDates, date => {
```
[the remark this answers is not in the document]
Comment (sauravmndl) [reply broker/lib/fabrik/DirectorManager.js]: done

File: api-controllers/ServiceFabrikApiController.js [hunk 20]
```diff
@@ -412,18 +412,18 @@ class ServiceFabrikApiController extends FabrikBaseController {
...
-          return Promise.try(() => this.validateRestoreTimeStamp(timeStamp))
-            .then(() => this.validateRestoreHistory({
-              instance_guid: instanceId,
-              service_id: serviceId,
-              plan_id: planId,
-              tenant_id: tenantId
-            }));
         }
       })
+      .then(() => this.validateRestoreHistory({
```
[the remark this answers is not in the document]
Comment (sauravmndl) [reply api-controllers/ServiceFabrikApiController.js]: done

File: api-controllers/ServiceFabrikApiController.js [hunk 21]
```diff
@@ -427,16 +434,30 @@ class ServiceFabrikApiController extends FabrikBaseController {
...
+        if (timeStamp) {
+          return this.backupStore
+            .listBackupsOlderThan(backupFileOptions, new Date(timeStamp))
+            .then(oldBackups =>
+              _.sortBy(oldBackups, ['started_at']))
+            .then(sortedOldBackups =>
+              _.findLast(sortedOldBackups, backup => backup.state ===  CONST.OPERATION.SUCCEEDED))
+            .tap(successfulBackup => {
+              if (_.isEmpty(successfulBackup)) {
+                logger.error(`No backup successful found for service instance '${instanceId}' before time_stamp ${new Date(timeStamp)}`);
```
[the remark this answers is not in the document]
Comment (sauravmndl) [reply api-controllers/ServiceFabrikApiController.js]: done

File: api-controllers/FabrikBaseController.js [hunk 22]
```diff
@@ -111,12 +113,43 @@ class FabrikBaseController extends BaseController {
...
+    return this.backupStore
+      .getRestoreFile(options)
+      .then(metdata => {
+        let restoreDates = _.get(metdata, 'restore_dates');
+        if (!_.isEmpty(restoreDates)) {
+          _.remove(restoreDates, date => {
+            const olderDateTillRestoreAllowed = Date.now() - 1000 * 60 * 60 * 24 * config.backup.restore_history_days;
+            return Date.parse(date) < olderDateTillRestoreAllowed;
+          });
+          //after removeing all older restore 'dateArray' conatains dates within allowed time
```
[the remark this answers is not in the document]
Comment (sauravmndl) [reply api-controllers/FabrikBaseController.js]: done

File: api-controllers/FabrikBaseController.js [hunk 23]
```diff
@@ -111,12 +113,43 @@ class FabrikBaseController extends BaseController {
...
+          // dates count should be less than 'config.backup.num_of_allowed_restores'
+          if (restoreDates.length >= config.backup.num_of_allowed_restores) {
+            throw new BadRequest(`Restore allowed only ${config.backup.num_of_allowed_restores} times within ${config.backup.restore_history_days} days.`);
+          }
+        }
+      })
+      .catch(NotFound, (err) => {
+        logger.debug('Not found any restore data.', err);
+        //Restoe file might not be found, first time restore.
```
[the remark this answers is not in the document]
Comment (sauravmndl) [reply api-controllers/FabrikBaseController.js]: done

File: api-controllers/ServiceFabrikApiController.js (hunk 17 above)
Comment (sauravmndl) [reply api-controllers/ServiceFabrikApiController.js]: Wanted to be independent at this part, so did once more. As this part  is critical for quota check. Also did in filtering old backups.

File: jobs/BnRStatusPollerJob.js (hunk 16 above)
Comment (pritishmishra) [reply jobs/BnRStatusPollerJob.js]: Added retry logic for schedule backup to retry even if the `scheduleBackup` call fails.

File: api-controllers/ServiceFabrikApiController.js (hunk 17 above)
Comment (jagadish-kb) [reply api-controllers/ServiceFabrikApiController.js]: why dont u just add a unit test to `listBackupsOlderThan` which always returns sorted in ascending that way if someone changes that the Unit test fails. These are within the same component so really dont need to be this independent & end up doing repetitive processing across files in the same code flow.

Comment (sauravmndl) [reply api-controllers/ServiceFabrikApiController.js]: ok..will do.

File: jobs/ScheduleBackupJob.js [hunk 24]
```diff
@@ -135,24 +137,58 @@ class ScheduleBackupJob extends BaseJob {
...
-          CONST.JOB.SCHEDULED_BACKUP,
-          RUN_AFTER,
-          jobData,
-          CONST.SYSTEM_USER
-        );
+      const plan = catalog.getPlan(jobData.plan_id);
+      return ScheduleManager.schedule(
+        jobData.instance_id,
+        CONST.JOB.SCHEDULED_BACKUP,
+        utils.getCronWithIntervalAndAfterXminute(plan.service.backup_interval, CONST.SCHEDULE.RETRY_DELAY),
```
Comment (sauravmndl) [inline jobs/ScheduleBackupJob.js]: can we make CONST.SCHEDULE.RETRY_DELAY as configurable ?

File: jobs/ScheduleBackupJob.js [hunk 25]
```diff
@@ -49,7 +50,7 @@ class ScheduleBackupJob extends BaseJob {
               .startBackup(_.pick(jobData, 'instance_id', 'type', 'trigger'))
               .catch(errors.Conflict, errors.UnprocessableEntity, (err) => {
                 logger.error('Some other operation already in progress on this instance:', err);
-                return retry(() => this.reScheduleBackup(job.attrs.data), {
+                return retry(() => this.reScheduleBackup(job.attrs.data, job.attrs.repeatInterval), {
```
Comment (ketakig12) [inline jobs/ScheduleBackupJob.js]: So, reScheduleBackup is done when backup fails with Conflict and UnprocessableEntity. When does UnprocessableEntity error comes and what about actual backup failure?

File: broker/lib/fabrik/DirectorManager.js [hunk 26]
```diff
@@ -871,18 +872,50 @@ class DirectorManager extends BaseManager {
...
+                  restore_dates: _.sortBy(restoreDates)
+                });
+            })
+            .tap(() => {
+              // Trigger schedule backup when restore is successful
+              if (lastOperation.state === CONST.OPERATION.SUCCEEDED) {
+                return this.reScheduleBackup({
+                  instance_id: options.instance_guid,
+                  afterXminute: config.backup.reschedule_backup_delay_after_restore || 3
+                });
```
[the remark this answers is not in the document]
Comment (jagadish-kb) [reply broker/lib/fabrik/DirectorManager.js]: last min comment, sorry missed in first round - should this scheduling of backup be done only for PITR restores? couple of ways to do it:
1. Have a feature flag named PITR and that is checked.
2. Check if any op logs are applied as part of restore (agent can return this info back as part of restore success status)

Based on the above you decide and reschedule accordingly. May be you keep this as-is in this PR, but add a JIRA to track this. So i will approve this PR with this item.

File: api-controllers/ServiceFabrikApiController.js [hunk 27]
```diff
@@ -481,6 +503,10 @@ class ServiceFabrikApiController extends FabrikBaseController {
   listBackups(req, res) {
     const options = _.pick(req.query, 'service_id', 'plan_id', 'instance_id', 'before', 'after');
     options.tenant_id = req.entity.tenant_id;
+    if (_.isEmpty(options.after)) {
+      const retentionPeriodOlderMillis = Date.now() - config.backup.retention_period_in_days * 24 * 60 * 60 * 1000;
+      options.after = new Date(retentionPeriodOlderMillis).toISOString();
+    }
```
[the remark this answers is not in the document]
Comment (sauravmndl) [reply api-controllers/ServiceFabrikApiController.js]: Yes, it should be handled by client which consumes sf API for backup list
